### DISCUSSION TITLE
What can awake sleeping bodies without presence of constraints?

### DISCUSSION THREAD

**Discussion by Frooxius** [bepu/bepuphysics2#139] [Q&A] [answered]
Hello!

I have another question again, this time I ran into a strange problem with my integration, I'm not too sure if it's a corner case or not, but I'd like to better understand what could be going on.

I have a number of kinematic bodies (a single Box that's a singular child of a Compound) that are densely overlapping each other and pretty large (hundreds of units). They are not being changed at all from my end, they have no velocity and their pose isn't being changed.

They're configured with BodyActivity description with 0.1f threshold and just 2 timesteps before they sleep. I've also verified following:
- NarrowPhaseCallbacks always return false in `AllowContactGeneration(int workerIndex, CollidableReference a, CollidableReference b)`. It gets called over 16000 times a tick.
- AllowContactGeneration(int workerIndex, CollidablePair pair, int childIndexA, int childIndexB) always returns true, but ends up  never being called in this setup)
- Their velocities stay at 0, TimestepsUnderThresholdCount keeps incrementing
- There are no constraints in the Solver - Solver.CountConstraints() returns 0 after a timestep

There's constantly around 63-64 active bodies (in the ActiveSet), Every frame some of them seem to sleep, but then another gets woken up again. It seems like it's a self perpetuating cycle, that prevents all of them from going to sleep, despite nothing changing externally.

I'm trying to understand what's going on. Why don't they all go to sleep? Since no contacts are allowed to be generated, what could be waking them up?

I've read up some of the info on Activity management here:
https://github.com/bepu/bepuphysics2/issues/21

I don't have best understanding of what exactly it does right now, but from what I can glean, is it possible they end up waking up because they become part of another island that is awake?

Any insights into what might be going on would definitely be appreciated. In this case, I plan on actually making them Statics instead, but I'm a bit confused by this behavior, but I'm trying to understand how things work under the hood so I avoid any performance gotcha's and such.

If it helps I can see if I can replicate this with the Demo project, if you'd like to do some testing, but I thought I'd ask first!

**Comment by RossNordby** [comment, marked as answer]
Islands are defined by the constraint connectivity graph, so in order for the bodies to be awoken by some other active bodies, there would need to be a constraint connecting them. With no contact constraints, it would have to be through non-contact constraints (e.g. BallSocket).

Apart from constraints, the only thing that wakes bodies up is calling functions which directly or indirectly wake them up. Anything that bottoms out calling the `IslandAwakener` functions `AwakenBody`, `AwakenConstraint`, `AwakenSet`, or `AwakenSets`.

`AwakenConstraint` is triggered when removing a constraint or calling `ApplyDescription` on a constraint. It will only wake bodies that are involved in the same island, so bodies with zero constraints can't be affected.

`AwakenBody` is used more widely. It will also wake up any body in the same constraint graph (island). The following will result in a call to it:
1. Not surprisingly, `BodyReference.Awake = true`
2. Removing a body from the simulation
3. `Bodies.SetLocalInertia` and its `BodyReference` wrapper
4. `Bodies.SetShape` and its `BodyReference` wrapper
5. `Bodies.ApplyDescription` and its `BodyReference` wrapper 
6. Adding a constraint
7. Removing a static from the simulation wakes up all bodies within its bounding box
8. `Statics.ApplyDescription` or its `StaticReference` wrapper wakes up all bodies within its bounding box
9. `Statics.SetShape` or its `StaticReference` wrapper wakes up all bodies within its bounding box
10.  Adding a static to the simulation wakes up all bodies within its bounding box

If I'm not mistaken, that's an exhaustive list as of 2.3.0/2.4.0. The unifying idea is that simulation modifications which would likely change what should be considered 'sleeping' should wake involved bodies up so that they can respond to the change. (And also it's sometimes just way simpler to do the involved bookkeeping when you know the involved bodies are active.)

**Reply by Frooxius** [reply]
Thank you so much! This makes a lot of sense, I was able to make adjustments so now the bodies stay properly asleep!

I didn't realize that certain changes and updates wake up all of the bodies in the bounding box, even if the bodies don't actually generate contacts. For the collider types that I know don't generate contacts, I just set the pose directly and only update broadphase manually, so it doesn't run the awakening.

As a more open ended question, would there be a good mechanism to customize this in the future? Say I make an update to a body that wakes up everything in the bounding box - could I filter which bodies in the bounding box to wake up and which not? Or would that be something that's better left to implementation - e.g. I do a broadphase query manually and wake up what I want there?

I don't need that right now (or maybe at all), but I'm just curious.

**Reply by RossNordby** [reply]
>would there be a good mechanism to customize this in the future? Say I make an update to a body that wakes up everything in the bounding box - could I filter which bodies in the bounding box to wake up and which not?

[perhaps...](https://github.com/bepu/bepuphysics2/commit/178cbd0766d9439f13e022cb0ec8d90495f29295)

**Reply by Frooxius** [reply]
Oh you're awesome! :D Thank you so much! I might actually end up using that instead since it seems like a cleaner way than doing it manually.
